scvi/external/cell2location/_model.py

```diff
@@ -0,0 +1,572 @@
+from typing import Optional, Tuple
+
+import numpy as np
+import pandas as pd
+import pyro
+from anndata import AnnData
+
+from scvi._compat import Literal
+from scvi.data import register_tensor_from_anndata
+from scvi.dataloaders import AnnDataLoader
+from scvi.lightning import PyroTrainingPlan, Trainer
+from scvi.model.base import BaseModelClass
+from scvi.external.cell2location._module import Cell2locationModule
+from pyro.infer import Predictive
+import torch
+
+from tqdm.auto import tqdm
+
+import matplotlib
+import matplotlib.pyplot as plt
+from scipy.optimize import linear_sum_assignment
+
+
+class Cell2locationBaseModelClass(BaseModelClass):
+    """
+    Reimplementation of cell2location [Kleshchevnikov20]_ model. BaseModelClass. 
+
+    https://github.com/BayraktarLab/cell2location
+
+    Parameters
+    ----------
+    sc_adata
+        single-cell AnnData object that has been registered via :func:`~scvi.data.setup_anndata`.
+    use_gpu
+        Use the GPU or not.
+    **model_kwargs
+        Keyword args for :class:`~scvi.external.cell2location...`
+
+    Examples
+    --------
+    >>> 
+    """
+
+    def __init__(
+        self,
+        adata: AnnData,
+        n_fact: int = 50,
+        var_names_read = None,
+        fact_names = None,
+        sample_id = None,
+        use_gpu: bool = True,
+        batch_size: int = 1024,
+        **model_kwargs,
+    ):
+        super(Cell2locationBaseModelClass, self).__init__(adata, use_gpu=use_gpu)
+        
+        adata.obs["_indices"] = np.arange(adata.n_obs).astype('int64')
+        register_tensor_from_anndata(adata, "ind_x", "obs", "_indices")
+        
+        self.adata = adata
+        
+        self.n_var = self.summary_stats["n_vars"]
+        self.n_obs = self.summary_stats["n_cells"]
+        self.n_fact = n_fact
+        self.batch_size = batch_size
+        
+        self.var_names = pd.Series(adata.var_names, index=adata.var_names)
+        if var_names_read is None:
+            self.var_names_read = pd.Series(self.var_names, index=self.var_names)
+        else:
+            self.var_names_read = pd.Series(adata.var[var_names_read], index=self.var_names)
+        self.obs_names = pd.Series(adata.obs_names, index=adata.obs_names)
+        
+        if fact_names is None:
+            self.fact_names = pd.Series(['fact_' + str(i) for i in range(self.n_fact)])
+        else:
+            self.fact_names = pd.Series(fact_names)
+            
+        if sample_id is None:
+            self.sample_id = pd.Series(['sample' for i in range(self.n_obs)],
+                                       index=self.obs_names)
+        else:
+            self.sample_id = pd.Series(adata.obs[sample_id], index=self.obs_names)
```

Comment:
I see. So you would do this on the fly?
```python
obs2sample = torch.nn.functional.one_hot(batch_key_index)
```
The alternative to this is whether pytorch tensor can be indexed like this:
```python
s_g_gene_add[batch_key_index,:]
```
rather than using one-hot encoded batch like this:
```python
torch.matmul(obs2sample, s_g_gene_add)
```

Is there a way to fetch the total number of batches/experiments (needed to create pyro plates)? (I find it a bit confusing to call both training batch and experiment batch as 'batch')